A30-xds-v3.md

```diff
@@ -0,0 +1,255 @@
+A30: xDS v3 Support
+----
+* Author(s): Mark D. Roth (markdroth)
+* Approver: ejona86, dfawley
+* Status: In Review
+* Implemented in: 
+* Last updated: 2020-06-25
+* Discussion at: https://groups.google.com/g/grpc-io/c/YRVqRMUQwhM
+
+## Abstract
+
+The purpose of this proposal is to describe how gRPC will handle the
+transition from v2 to v3 of the xDS API and to document how gRPC will handle
+xDS versioning on an ongoing basis.
+
+## Background
+
+gRPC has recently added support for obtaining configuration via the
+[xDS API](https://github.com/envoyproxy/data-plane-api).  Our current
+implementation uses uses v2 of the xDS API, but we are now adding
+support for [xDS
+v3](https://docs.google.com/document/d/1xeVvJ6KjFBkNjVspPbY_PwEDHC7XPi0J5p1SqUXcCl8/edit).
+In addition, we have been working with the xDS community on the new
+[minor/patch versioning
+proposal](https://docs.google.com/document/d/1afQ9wthJxofwOMCAle8qF-ckEfNWUuWtxbmdhxji0sI/edit),
+which should provide a smoother transition mechanism going forward.
+
+Regardless of whether a new version is indicated by bumping the
+major version (as was done with the v2 to v3 transition) or by bumping
+the minor version (as will probably be done going forward), the current
+plan is that there will still be one new version of the xDS API per
+year.
+
+### Related Proposals: 
+
+[A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md)
+
+## Proposal
+
+### Version Support Policy
+
+gRPC's policy will be to support two versions of xDS at any given time.
+So if gRPC currently supports versions N-1 and N, then when xDS version
+N+1 is released, gRPC will add support for version N+1 and drop support
+for version N-1 at the same time, so that we will support versions N
+and N+1.
+
+This policy will give users 1-2 years to upgrade their xDS servers in
+order to continue using the latest release of gRPC.  Older releases of
+gRPC will of course continue to function using the older versions of the
+xDS API.
+
+### Deciding What Version to Use
+
+Any given gRPC release will support two versions of xDS, versions N-1 and N.
+It will assume that all servers speak version N-1, and there will be a
```

Comment:
You were the one who suggested this mechanism. :)  The main benefit of expressing this as a sever capability instead of just configuring what version to use is that in the future, it can allow the same bootstrap file to work for both a client that supports versions [N-1, N] and a client that supports versions [N, N+1].

I think we can support a good error message even with this design.  We know what versions we think the server supports, and we know what version we decided to try, so we can put both of those things in the error message if we get an UNIMPLEMENTED status.

Reply:
I'm not arguing against the mechanism. The server capability works. I'm arguing against the "assume."

> We know what versions we think the server supports

No we don't. N-1 doesn't need to be included at all. I want N-1 to be included and us not to assume it. I'm fine with assuming v2, for backward compatibility, but once we support v3.1 we should only care about explicitly specified versions.

> so we can put both of those things in the error message if we get an UNIMPLEMENTED status

In later minor versions, we won't get UNIMPLEMENTED status, because the package didn't change.

Reply:
I see, so you'd prefer an approach where, when we add support for v3.1, gRPC always explicitly looks for a server feature for both v3 and v3.1, and doesn't automatically assume v3?

I'm not actually sure this makes much practical difference, but I guess it is a little easier to explain.  I've updated the doc accordingly.  PTAL.